llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -883,6 +881,110 @@ static bool removeDeviceGlobalFromCompilerUsed(Module &M) {
   return true;
 }
 
+// Add extra red zone to each image scope device globals if the module has been
+// instrumented by sanitizer pass. And record their infomation like size, red
+// zone size, beginning address.
+static bool instrumentDeviceGlobal(Module &M) {
+  auto &DL = M.getDataLayout();
+  IRBuilder<> IRB(M.getContext());
+  SmallVector<GlobalVariable *, 8> GlobalsToRemove;
+  SmallVector<GlobalVariable *, 8> NewDeviceGlobals;
+  SmallVector<Constant *, 8> DeviceGlobalMetadata;
+
+  constexpr uint64_t MaxRZ = 1 << 18;
+  const uint64_t MinRZ = 32;
+
+  Type *IntptrTy =
+      Type::getIntNTy(M.getContext(), M.getDataLayout().getPointerSizeInBits());
+
+  // Device global meta data is described by a structure
+  //  size_t device_global_size
+  //  size_t device_global_size_with_red_zone
+  //  size_t beginning address of the device global
+  StructType *StructTy = StructType::get(IntptrTy, IntptrTy, IntptrTy);
+
+  for (auto &G : M.globals()) {
+    // Non image scope device globals are implemented by device USM, and the
+    // out-of-bounds check for them will be done by sanitizer USM part. So we
+    // exclude them here.
+    if (isDeviceGlobalVariable(G) && hasDeviceImageScopeProperty(G)) {
+      Type *Ty = G.getValueType();
+      const uint64_t SizeInBytes = DL.getTypeAllocSize(Ty);
+      const uint64_t RightRedzoneSize = [&] {
+        uint64_t RZ = 0;
+        if (SizeInBytes <= MinRZ / 2) {
+          // Reduce redzone size for small size objects, e.g. int, char[1].
+          // Optimize when SizeInBytes is less than or equal to half of MinRZ.
+          RZ = MinRZ - SizeInBytes;
+        } else {
+          // Calculate RZ, where MinRZ <= RZ <= MaxRZ, and RZ ~ 1/4 *
+          // SizeInBytes.
+          RZ = std::clamp((SizeInBytes / MinRZ / 4) * MinRZ, MinRZ, MaxRZ);
```

Comment:
Am I right that `(SizeInBytes / MinRZ / 4) * MinRZ` is written this way on purpose, to ensure certain rounding during calculation? If so, a comment would be welcome about that.

Reply:
This code is coming from https://github.com/intel/llvm/blob/sycl/llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp#L2767.
I traced back the git histories. It seems that this code initially started out like this. So, I also don't know if it's written this way for some reason. Sorry for this.
But we do this instrumentation for device global in sycl-post-link tool temporarily. When unified runtime implemented the API urProgramGetGlobalVariablePointer, I'll move this instrumentation code back to AddressSanitizer pass.